src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs

```diff
@@ -97,11 +97,22 @@ private static async Task CacheSemanticClassificationsAsync(Document document, C
                 return;
 
             using var _2 = ArrayBuilder<ClassifiedSpan>.GetInstance(out var classifiedSpans);
-
-            // Compute classifications for the full span.
             var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-            var options = ClassificationOptions.From(document.Project);
-            await classificationService.AddSemanticClassificationsAsync(document, new TextSpan(0, text.Length), options, classifiedSpans, cancellationToken).ConfigureAwait(false);
+
+            // Compute classifications for the full span. Razor documents need to cache both syntactic and semantic
+            // classifications since the client doesn't run the C# classifier in Razor LSP scenarios.
+            // Ideally, Razor will eventually run the C# syntactic classifier on their end so we can remove this
+            // special casing: https://github.com/dotnet/razor-tooling/issues/5850
+            if (document.FilePath is not null && document.FilePath.EndsWith(".g.cs"))
```

Comment:
`.IsRazorDocument()` doesn't work in OOP scenarios, so as a workaround I added a check for `.EndsWith(".g.cs")` instead which seems to do the trick

Reply:
@NTaylorMullen for thoughts.

Reply:
note: This feels weird as well.  somethign about the entire razor space here seems somewhat broken (necessitating these extra strange codepaths at the feature level).  my preference would be that if razor is *currently* doing something different/weird, that we just not have this optimization/feature for them.  and, if we can then get razor docs/scenarios to work like normal scenarios, then we reenable, with no specialized code *in the feature* to handle razor.

Reply:
Most of this special casing is due to Razor not running the C# syntax classifier on their end. However, I believe adding support is a pretty big work item and could possibly also negatively impact Razor perf, so the effects would have to be investigated.

While this special casing doesn't look good, I think it will result in pretty good gains for Razor semantic tokens which imo might be an appropriate trade-off since the Razor LSP editor is on by default for users.